api/src/apps/apps.rs

```diff
@@ -285,6 +283,46 @@ impl AppsService {
         Ok(services)
     }
 
+    /// Adds all companions to service_configs that are not yet present. Also updates
+    /// service configs if there is a corresponding companion config.
+    fn merge_companions_by_service_name(
+        companions: Vec<ServiceConfig>,
+        service_configs: &mut Vec<ServiceConfig>,
+    ) -> () {
+        for service in service_configs.iter_mut() {
+            let matching_companion = companions
+                .iter()
+                .find(|companion| companion.service_name() == service.service_name());
+
+            if matching_companion.is_some() {
+                AppsService::merge_service_configs(service, matching_companion.unwrap());
+            }
+        }
+
+        service_configs.extend(AppsService::filter_companions_by_service_name(
+            companions,
+            service_configs,
+        ));
+    }
+
+    fn merge_service_configs(to: &mut ServiceConfig, from: &ServiceConfig) -> () {
```

Comment:
In an other branch I started to put more logic into the `ServiceConfig` itself because I try to avoid to implement anemic domain models. Therefore, I would create a method in `ServiceConfig`. For example:

```rust
impl ServiceConfig {
    pub fn merge_with(&mut self, &other: Self) {
        // ...
    }
}
```

This provides the ability to implement the merge tests there.